Error while checking build/O2/fullCI for ada0fcb39be1e465dbb3cc012795b9b118d848bf at 2023-01-19 08:26:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10593-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10593/ada0fcb39be1e465dbb3cc012795b9b118d848bf/build_O2_fullCI/pretty.html).

--- a/DataFormats/Detectors/FOCAL/include/DataFormatsFOCAL/PixelChipRecord.h
+++ b/DataFormats/Detectors/FOCAL/include/DataFormatsFOCAL/PixelChipRecord.h
@@ -37,7 +37,7 @@ class PixelChipRecord
   int getLaneID() const { return mLaneID; }
   int getChipID() const { return mChipID; }
   int getFirstHit() const { return mHitIndexRange.getFirstEntry(); }
-  int getNumberOfHits() const { return mHitIndexRange.getFirstEntry(); }
+  int getNumberOfHits() const { return mHitIndexRange.getEntries(); }
 
   void printStream(std::ostream& stream) const;
 

--- a/DataFormats/Detectors/FOCAL/src/Event.cxx
+++ b/DataFormats/Detectors/FOCAL/src/Event.cxx
@@ -11,6 +11,8 @@
 #include <algorithm>
 #include <DataFormatsFOCAL/ErrorHandling.h>
 #include <DataFormatsFOCAL/Event.h>
+#include <iostream>
+
 using namespace o2::focal;
 
 PadLayerEvent& Event::getPadLayer(unsigned int index)
@@ -69,7 +71,21 @@ void Event::construct(const o2::InteractionRecord& interaction, gsl::span<const 
   }
 
   for (auto& chip : eventPixels) {
-    mPixelLayers[chip.getLayerID()].addChip(chip.getLaneID(), chip.getChipID(), pixelHits.subspan(chip.getFirstHit(), chip.getNumberOfHits()));
+    if (chip.getLayerID() > 1) {
+      std::cerr << "Invalid layer ID chip " << chip.getChipID() << ": " << chip.getLayerID() << std::endl;
+      continue;
+    }
+    if (chip.getNumberOfHits()) {
+      if (chip.getFirstHit() >= pixelHits.size()) {
+        std::cerr << "First hit index " << chip.getFirstHit() << " exceeding hit contiainer " << pixelHits.size() << std::endl;
+        continue;
+      }
+      if (chip.getFirstHit() + chip.getNumberOfHits() - 1 >= pixelHits.size()) {
+        std::cerr << "First hit index " << chip.getFirstHit() + chip.getNumberOfHits() - 1 << " exceeding hit contiainer " << pixelHits.size() << std::endl;
+        continue;
+      }
+      mPixelLayers[chip.getLayerID()].addChip(chip.getLaneID(), chip.getChipID(), pixelHits.subspan(chip.getFirstHit(), chip.getNumberOfHits()));
+    }
   }
 }
 

--- a/Detectors/FOCAL/base/include/FOCALBase/EventReader.h
+++ b/Detectors/FOCAL/base/include/FOCALBase/EventReader.h
@@ -44,6 +44,8 @@ class EventReader
   std::unique_ptr<TTreeReaderValue<std::vector<PixelHit>>> mPixelHitBranch;
   std::unique_ptr<TTreeReaderValue<std::vector<PixelChipRecord>>> mPixelChipBranch;
   std::unique_ptr<TTreeReaderValue<std::vector<TriggerRecord>>> mTriggerBranch;
+
+  ClassDefNV(EventReader, 1);
 };
 
 } // namespace o2::focal

--- a/Detectors/FOCAL/workflow/include/FOCALWorkflow/RawDecoderSpec.h
+++ b/Detectors/FOCAL/workflow/include/FOCALWorkflow/RawDecoderSpec.h
@@ -42,6 +42,7 @@ class RawDecoderSpec : public framework::Task
     std::vector<std::array<PadLayerEvent, constants::PADS_NLAYERS>> mPadEvents;
     std::vector<std::array<PixelLayerEvent, constants::PIXELS_NLAYERS>> mPixelEvent;
     std::vector<o2::InteractionRecord> mPixelTriggers;
+    std::vector<std::vector<int>> mFEEs;
   };
   RawDecoderSpec() = default;
   RawDecoderSpec(uint32_t outputSubspec, bool usePadData, bool usePixelData, bool debug) : mDebugMode(debug), mUsePadData(usePadData), mUsePixelData(usePixelData), mOutputSubspec(outputSubspec) {}
@@ -56,12 +57,13 @@ class RawDecoderSpec : public framework::Task
  private:
   void sendOutput(framework::ProcessingContext& ctx);
   void resetContainers();
-  int decodePadData(const gsl::span<const char> padWords, o2::InteractionRecord& interaction);
-  void decodePadEvent(const gsl::span<const char> padWords, o2::InteractionRecord& interaction);
-  int decodePixelData(const gsl::span<const char> pixelWords, o2::InteractionRecord& interaction, int fecID);
+  int decodePadData(const gsl::span<const char> padWords, o2::InteractionRecord& hbIR);
+  void decodePadEvent(const gsl::span<const char> padWords, o2::InteractionRecord& hbIR);
+  int decodePixelData(const gsl::span<const char> pixelWords, o2::InteractionRecord& hbIR, int fecID);
   std::array<PadLayerEvent, constants::PADS_NLAYERS> createPadLayerEvent(const o2::focal::PadData& data) const;
   void fillChipsToLayer(PixelLayerEvent& pixellayer, const gsl::span<const PixelChip>& chipData);
   void fillEventPixeHitContainer(std::vector<PixelHit>& eventHits, std::vector<PixelChipRecord>& eventChips, const PixelLayerEvent& pixelLayer, int layerIndex);
+  int filterIncompletePixelsEventsHBF(HBFData& data, const std::vector<int>& expectFEEs);
   void buildEvents();
   bool consistencyCheckPixelFEE(const std::unordered_map<int, int>& counters) const;
   int maxCounter(const std::unordered_map<int, int>& counters) const;
@@ -73,6 +75,7 @@ class RawDecoderSpec : public framework::Task
   bool mDebugMode = false;
   bool mUsePadData = true;
   bool mUsePixelData = true;
+  bool mFilterIncomplete = false;
   uint32_t mOutputSubspec = 0;
   PadDecoder mPadDecoder;
   PixelDecoder mPixelDecoder;

--- a/Detectors/FOCAL/workflow/src/RawDecoderSpec.cxx
+++ b/Detectors/FOCAL/workflow/src/RawDecoderSpec.cxx
@@ -29,6 +29,10 @@ using namespace o2::focal::reco_workflow;
 
 void RawDecoderSpec::init(framework::InitContext& ctx)
 {
+  if (ctx.options().get<bool>("filterIncomplete")) {
+    LOG(info) << "Enabling filtering of incomplete events in the pixel data";
+    mFilterIncomplete = true;
+  }
 }
 
 void RawDecoderSpec::run(framework::ProcessingContext& ctx)
@@ -39,11 +43,11 @@ void RawDecoderSpec::run(framework::ProcessingContext& ctx)
   int inputs = 0;
   std::vector<char> rawbuffer;
   int currentendpoint = 0;
-  bool nextHBF = true;
   o2::InteractionRecord currentIR;
   std::unordered_map<int, int> numHBFFEE, numEventsFEE;
   std::unordered_map<int, std::vector<int>> numEventsHBFFEE;
   int numHBFPadsTF = 0, numEventsPadsTF = 0;
+  std::vector<int> expectFEEs;
   for (const auto& rawData : framework::InputRecordWalker(ctx.inputs())) {
     if (rawData.header != nullptr && rawData.payload != nullptr) {
       const auto payloadSize = o2::framework::DataRefUtils::getPayloadSize(rawData);
@@ -52,85 +56,83 @@ void RawDecoderSpec::run(framework::ProcessingContext& ctx)
 
       gsl::span<const char> databuffer(rawData.payload, payloadSize);
       int currentpos = 0;
+      bool firstHBF = true;
       while (currentpos < databuffer.size()) {
         auto rdh = reinterpret_cast<const o2::header::RDHAny*>(databuffer.data() + currentpos);
         if (mDebugMode) {
           o2::raw::RDHUtils::printRDH(rdh);
         }
-        if (o2::raw::RDHUtils::getMemorySize(rdh) == o2::raw::RDHUtils::getHeaderSize(rdh)) {
-          auto trigger = o2::raw::RDHUtils::getTriggerType(rdh);
-          if (trigger & o2::trigger::SOT || trigger & o2::trigger::HB) {
-            if (o2::raw::RDHUtils::getStop(rdh)) {
-              LOG(debug) << "Stop bit received - processing payload";
-              // Data ready
-              if (currentendpoint == 1) {
-                // Pad data
-                if (mUsePadData) {
-                  LOG(debug) << "Processing PAD data";
-                  auto nEventsPads = decodePadData(rawbuffer, currentIR);
-                  auto found = mNumEventsHBFPads.find(nEventsPads);
-                  if (found != mNumEventsHBFPads.end()) {
-                    found->second += 1;
-                  } else {
-                    mNumEventsHBFPads.insert(std::pair<int, int>{nEventsPads, 1});
-                  }
-                  numEventsPadsTF += nEventsPads;
-                  numHBFPadsTF++;
+        auto trigger = o2::raw::RDHUtils::getTriggerType(rdh);
+        if (trigger & o2::trigger::HB) {
+          // HB trigger received
+          if (o2::raw::RDHUtils::getStop(rdh)) {
+            LOG(debug) << "Stop bit received - processing payload";
+            // Data ready
+            if (currentendpoint == 1) {
+              // Pad data
+              if (mUsePadData) {
+                LOG(debug) << "Processing PAD data";
+                auto nEventsPads = decodePadData(rawbuffer, currentIR);
+                auto found = mNumEventsHBFPads.find(nEventsPads);
+                if (found != mNumEventsHBFPads.end()) {
+                  found->second += 1;
+                } else {
+                  mNumEventsHBFPads.insert(std::pair<int, int>{nEventsPads, 1});
                 }
-              } else if (currentendpoint == 0) {
-                // Pixel data
-                if (mUsePixelData) {
-                  auto feeID = o2::raw::RDHUtils::getFEEID(rdh);
-                  LOG(debug) << "Processing Pixel data from FEE " << feeID;
-                  auto neventsPixels = decodePixelData(rawbuffer, currentIR, feeID);
-                  auto found = numHBFFEE.find(feeID);
-                  if (found != numHBFFEE.end()) {
-                    found->second++;
-                  } else {
-                    numHBFFEE.insert(std::pair<int, int>{feeID, 1});
-                  }
-                  auto evFound = numEventsFEE.find(feeID);
-                  if (evFound != numEventsFEE.end()) {
-                    evFound->second += neventsPixels;
-                  } else {
-                    numEventsFEE.insert(std::pair<int, int>{feeID, neventsPixels});
-                  }
-                  auto evHBFFound = numEventsHBFFEE.find(feeID);
-                  if (evHBFFound != numEventsHBFFEE.end()) {
-                    evHBFFound->second.push_back(neventsPixels);
-                  } else {
-                    std::vector<int> evbuffer;
-                    evbuffer.push_back(neventsPixels);
-                    numEventsHBFFEE.insert(std::pair<int, std::vector<int>>{feeID, evbuffer});
-                  }
-                }
-              } else {
-                LOG(error) << "Unsupported endpoint " << currentendpoint;
+                numEventsPadsTF += nEventsPads;
+                numHBFPadsTF++;
               }
-              nextHBF = true;
+            } else if (currentendpoint == 0) {
+              // Pixel data
+              if (mUsePixelData) {
+                auto feeID = o2::raw::RDHUtils::getFEEID(rdh);
+                if (firstHBF) {
+                  auto found = std::find(expectFEEs.begin(), expectFEEs.end(), feeID);
+                  if (found == expectFEEs.end()) {
+                    expectFEEs.emplace_back(feeID);
+                  }
+                  firstHBF = false;
+                }
+                LOG(debug) << "Processing Pixel data from FEE " << feeID;
+                auto neventsPixels = decodePixelData(rawbuffer, currentIR, feeID);
+                auto found = numHBFFEE.find(feeID);
+                if (found != numHBFFEE.end()) {
+                  found->second++;
+                } else {
+                  numHBFFEE.insert(std::pair<int, int>{feeID, 1});
+                }
+                auto evFound = numEventsFEE.find(feeID);
+                if (evFound != numEventsFEE.end()) {
+                  evFound->second += neventsPixels;
+                } else {
+                  numEventsFEE.insert(std::pair<int, int>{feeID, neventsPixels});
+                }
+                auto evHBFFound = numEventsHBFFEE.find(feeID);
+                if (evHBFFound != numEventsHBFFEE.end()) {
+                  evHBFFound->second.push_back(neventsPixels);
+                } else {
+                  std::vector<int> evbuffer;
+                  evbuffer.push_back(neventsPixels);
+                  numEventsHBFFEE.insert(std::pair<int, std::vector<int>>{feeID, evbuffer});
+                }
+              }
             } else {
-              LOG(debug) << "New HBF or Timeframe";
-              nextHBF = true;
+              LOG(error) << "Unsupported endpoint " << currentendpoint;
             }
+          } else {
+            rawbuffer.clear();
+            // Get interaction record for HBF
+            currentIR.bc = o2::raw::RDHUtils::getTriggerBC(rdh);
+            currentIR.orbit = o2::raw::RDHUtils::getTriggerOrbit(rdh);
+            currentendpoint = o2::raw::RDHUtils::getEndPointID(rdh);
+            LOG(debug) << "New HBF " << currentIR.orbit << " / " << currentIR.bc << ", endpoint " << currentendpoint;
           }
+        }
+
+        if (o2::raw::RDHUtils::getMemorySize(rdh) == o2::raw::RDHUtils::getHeaderSize(rdh)) {
+          // Skip page if emtpy
           currentpos += o2::raw::RDHUtils::getOffsetToNext(rdh);
           continue;
-        }
-
-        auto trigger = o2::raw::RDHUtils::getTriggerType(rdh);
-        if (trigger & o2::trigger::SOT || trigger & o2::trigger::HB) {
-          // HBF trigger could be part of regular payoad RDH (pixels)
-          nextHBF = true;
-        }
-
-        if (nextHBF) {
-          rawbuffer.clear();
-          // Get interaction record for HBF
-          currentIR.bc = o2::raw::RDHUtils::getTriggerBC(rdh);
-          currentIR.orbit = o2::raw::RDHUtils::getTriggerOrbit(rdh);
-          currentendpoint = o2::raw::RDHUtils::getEndPointID(rdh);
-          LOG(debug) << "New HBF " << currentIR.orbit << " / " << currentIR.bc << ", endpoint " << currentendpoint;
-          nextHBF = false;
         }
 
         // non-0 payload size:
@@ -175,8 +177,15 @@ void RawDecoderSpec::run(framework::ProcessingContext& ctx)
   }
   fillPixelEventHBFCount(numEventsHBFFEE);
 
-  LOG(info)
-    << "Found " << mHBFs.size() << " HBFs in timeframe";
+  if (mUsePixelData && mFilterIncomplete) {
+    std::cout << "Filtering incomplete pixel events" << std::endl;
+    for (auto& hbf : mHBFs) {
+      auto numErased = filterIncompletePixelsEventsHBF(hbf.second, expectFEEs);
+      mNumEventsPixels -= numErased;
+    }
+  }
+
+  LOG(info) << "Found " << mHBFs.size() << " HBFs in timeframe";
 
   buildEvents();
 
@@ -239,36 +248,36 @@ void RawDecoderSpec::resetContainers()
   mOutputTriggerRecords.clear();
 }
 
-int RawDecoderSpec::decodePadData(const gsl::span<const char> padWords, o2::InteractionRecord& interaction)
+int RawDecoderSpec::decodePadData(const gsl::span<const char> padWords, o2::InteractionRecord& hbIR)
 {
-  LOG(debug) << "Decoding pad data for Orbit " << interaction.orbit << ", BC " << interaction.bc;
+  LOG(debug) << "Decoding pad data for Orbit " << hbIR.orbit << ", BC " << hbIR.bc;
   constexpr std::size_t EVENTSIZEPADGBT = 1180,
                         EVENTSIZECHAR = EVENTSIZEPADGBT * sizeof(PadGBTWord) / sizeof(char);
   auto nevents = padWords.size() / (EVENTSIZECHAR);
   for (int ievent = 0; ievent < nevents; ievent++) {
-    decodePadEvent(padWords.subspan(EVENTSIZECHAR * ievent, EVENTSIZECHAR), interaction);
+    decodePadEvent(padWords.subspan(EVENTSIZECHAR * ievent, EVENTSIZECHAR), hbIR);
   }
   return nevents;
 }
 
-void RawDecoderSpec::decodePadEvent(const gsl::span<const char> padWords, o2::InteractionRecord& interaction)
+void RawDecoderSpec::decodePadEvent(const gsl::span<const char> padWords, o2::InteractionRecord& hbIR)
 {
   gsl::span<const PadGBTWord> padWordsGBT(reinterpret_cast<const PadGBTWord*>(padWords.data()), padWords.size() / sizeof(PadGBTWord));
   mPadDecoder.reset();
   mPadDecoder.decodeEvent(padWordsGBT);
-  std::map<o2::InteractionRecord, HBFData>::iterator foundHBF = mHBFs.find(interaction);
+  std::map<o2::InteractionRecord, HBFData>::iterator foundHBF = mHBFs.find(hbIR);
   if (foundHBF == mHBFs.end()) {
     // New event, create new entry
     HBFData nexthbf;
-    auto res = mHBFs.insert({interaction, nexthbf});
+    auto res = mHBFs.insert({hbIR, nexthbf});
     foundHBF = res.first;
   }
   foundHBF->second.mPadEvents.push_back(createPadLayerEvent(mPadDecoder.getData()));
 }
 
-int RawDecoderSpec::decodePixelData(const gsl::span<const char> pixelWords, o2::InteractionRecord& interaction, int feeID)
+int RawDecoderSpec::decodePixelData(const gsl::span<const char> pixelWords, o2::InteractionRecord& hbIR, int feeID)
 {
-  LOG(debug) << "Decoding pixel data for Orbit " << interaction.orbit << ", BC " << interaction.bc;
+  LOG(debug) << "Decoding pixel data for Orbit " << hbIR.orbit << ", BC " << hbIR.bc;
   auto fee = feeID & 0x00FF,
        branch = (feeID & 0x0F00) >> 8;
   int layer = fee < 2 ? 0 : 1;
@@ -282,13 +291,17 @@ int RawDecoderSpec::decodePixelData(const gsl::span<const char> pixelWords, o2::
   int nevents = 0;
   for (auto& [trigger, chipdata] : mPixelDecoder.getChipData()) {
     LOG(debug) << "Found trigger orbit " << trigger.orbit << ", BC " << trigger.bc;
+    if (trigger.orbit != hbIR.orbit) {
+      LOG(debug) << "FEE 0x" << std::hex << feeID << std::dec << ": Discarding spurious trigger with Orbit " << trigger.orbit << " (HB " << hbIR.orbit << ")";
+      continue;
+    }
     if (foundHBF == mHBFs.end()) {
       // take HBF from the first trigger as BC in RDH for pixel data is unreliable
-      foundHBF = mHBFs.find(interaction);
+      foundHBF = mHBFs.find(hbIR);
       if (foundHBF == mHBFs.end()) {
         // New event, create new entry
         HBFData nexthbf;
-        auto res = mHBFs.insert({trigger, nexthbf});
+        auto res = mHBFs.insert({hbIR, nexthbf});
         foundHBF = res.first;
       }
     }
@@ -298,6 +311,7 @@ int RawDecoderSpec::decodePixelData(const gsl::span<const char> pixelWords, o2::
       //
       auto index = triggerfound - foundHBF->second.mPixelTriggers.begin();
       fillChipsToLayer(foundHBF->second.mPixelEvent[index][layer], chipdata);
+      foundHBF->second.mFEEs[index].push_back(feeID);
     } else {
       // new trigger
       std::array<PixelLayerEvent, constants::PIXELS_NLAYERS> nextevent;
@@ -305,6 +319,7 @@ int RawDecoderSpec::decodePixelData(const gsl::span<const char> pixelWords, o2::
       foundHBF->second.mPixelTriggers.push_back(trigger);
       auto& current = foundHBF->second.mPixelEvent.back();
       fillChipsToLayer(current[layer], chipdata);
+      foundHBF->second.mFEEs.push_back({feeID});
     }
     nevents++;
   }
@@ -360,7 +375,7 @@ void RawDecoderSpec::buildEvents()
       LOG(debug) << "Processing HBF with IR: " << hbir.orbit << " / " << hbir.bc << std::endl;
       // check consistency in number of events between triggers, pixels and pads
       // in case all events are in the stream
-      if (!(hbf.mPadEvents.size() == hbf.mPixelEvent.size() == hbf.mPixelTriggers.size())) {
+      if ((hbf.mPadEvents.size() != hbf.mPixelEvent.size()) || (hbf.mPadEvents.size() != hbf.mPixelTriggers.size()) || (hbf.mPixelEvent.size() != hbf.mPixelTriggers.size())) {
         LOG(error) << "Inconsistent number of events in HBF for pads (" << hbf.mPadEvents.size() << ") and pixels (" << hbf.mPixelEvent.size() << ") - " << hbf.mPixelTriggers.size() << " triggers";
         continue;
       }
@@ -379,7 +394,7 @@ void RawDecoderSpec::buildEvents()
         }
         std::copy(eventHits.begin(), eventHits.end(), std::back_inserter(mOutputPixelHits));
         std::copy(eventPixels.begin(), eventPixels.end(), std::back_inserter(mOutputPixelChips));
-        mOutputTriggerRecords.emplace_back(hbf.mPixelTriggers[itrg], startPads, constants::PADS_NLAYERS, startHits, eventPixels.size(), startHits, eventHits.size());
+        mOutputTriggerRecords.emplace_back(hbf.mPixelTriggers[itrg], startPads, constants::PADS_NLAYERS, startChips, eventPixels.size(), startHits, eventHits.size());
       }
     } else if (mUsePixelData) {
       // only pixel data available, merge pixel layers and interaction record
@@ -416,6 +431,51 @@ void RawDecoderSpec::buildEvents()
       }
     }
   }
+}
+
+int RawDecoderSpec::filterIncompletePixelsEventsHBF(HBFData& data, const std::vector<int>& expectFEEs)
+{
+  auto same = [](const std::vector<int>& lhs, const std::vector<int>& rhs) -> bool {
+    bool missing = false;
+    for (auto entry : lhs) {
+      if (std::find(rhs.begin(), rhs.end(), entry) == rhs.end()) {
+        missing = true;
+        break;
+      }
+    }
+    if (!missing) {
+      for (auto entry : rhs) {
+        if (std::find(lhs.begin(), lhs.end(), entry) == lhs.end()) {
+          missing = true;
+          break;
+        }
+      }
+    }
+    return missing;
+  };
+  std::vector<int> indexIncomplete;
+  for (auto index = 0; index < data.mFEEs.size(); index++) {
+    if (data.mFEEs[index].size() != expectFEEs.size()) {
+      indexIncomplete.emplace_back(index);
+      continue;
+    }
+    if (!same(data.mFEEs[index], expectFEEs)) {
+      indexIncomplete.emplace_back(index);
+    }
+  }
+  if (indexIncomplete.size()) {
+    std::sort(indexIncomplete.begin(), indexIncomplete.end(), std::less<>());
+    // start removing from the end, since erase will impact the indexing
+    for (auto indexIter = indexIncomplete.rbegin(); indexIter != indexIncomplete.rend(); indexIter++) {
+      auto iterPixelEvent = data.mPixelEvent.begin() + *indexIter;
+      auto iterTrigger = data.mPixelTriggers.begin() + *indexIter;
+      auto iterFEEs = data.mFEEs.begin() + *indexIter;
+      data.mPixelEvent.erase(iterPixelEvent);
+      data.mPixelTriggers.erase(iterTrigger);
+      data.mFEEs.erase(iterFEEs);
+    }
+  }
+  return indexIncomplete.size();
 }
 
 bool RawDecoderSpec::consistencyCheckPixelFEE(const std::unordered_map<int, int>& counters) const
@@ -469,7 +529,7 @@ int RawDecoderSpec::maxCounter(const std::unordered_map<int, int>& counters) con
 void RawDecoderSpec::printCounters(const std::unordered_map<int, int>& counters) const
 {
   for (auto& [fee, counter] : counters) {
-    std::cout << "  FEE " << fee << ": " << counter << " counts ..." << std::endl;
+    std::cout << "  FEE 0x" << std::hex << fee << std::dec << ": " << counter << " counts ..." << std::endl;
   }
 }
 
@@ -486,7 +546,7 @@ void RawDecoderSpec::printEvents(const std::unordered_map<int, std::vector<int>>
       }
       stringbuilder << ev;
     }
-    std::cout << "  FEE " << fee << ": " << stringbuilder.str() << " events ..." << std::endl;
+    std::cout << "  FEE 0x" << std::hex << fee << std::dec << ": " << stringbuilder.str() << " events ..." << std::endl;
   }
 }
 
@@ -537,5 +597,6 @@ o2::framework::DataProcessorSpec o2::focal::reco_workflow::getRawDecoderSpec(boo
                                           inputs,
                                           outputs,
                                           o2::framework::adaptFromTask<o2::focal::reco_workflow::RawDecoderSpec>(outputSubspec, usePadData, usePixelData, debugMode),
-                                          o2::framework::Options{}};
+                                          o2::framework::Options{
+                                            {"filterIncomplete", o2::framework::VariantType::Bool, false, {"Filter incomplete pixel events"}}}};
 }

